jooqdb/src/main/java/org/cristalise/storage/jooqdb/JooqClusterStorage.java

```diff
@@ -1,339 +1,426 @@
+    }
+
+    @Override
+    public ClusterType[] getClusters(ItemPath itemPath) throws PersistencyException {
+        ArrayList<ClusterType> result = new ArrayList<ClusterType>();
+ 
+        for (ClusterType type:jooqHandlers.keySet()) {
+            if (jooqHandlers.get(type).exists(JooqHandler.connect(), itemPath.getUUID())) result.add(type);
+        }
+
+        return result.toArray(new ClusterType[0]);
+    }
+
+    @Override
+    public String[] getClusterContents(ItemPath itemPath, String path) throws PersistencyException {
+        if (StringUtils.isBlank(path)) {
+            ArrayList<String> result = new ArrayList<String>();
+
+            for (ClusterType k: getClusters(itemPath)) { result.add(k.toString()); }
+
+            return result.toArray(new String[0]);
+        }
+
+        UUID uuid = itemPath.getUUID();
+
+        String[]    pathArray   = path.split("/");
+        String[]    primaryKeys = Arrays.copyOfRange(pathArray, 1, pathArray.length);
+        ClusterType cluster     = ClusterType.getValue(pathArray[0]);
+
+        JooqHandler handler = jooqHandlers.get(cluster);
+
+        if (handler != null) {
+            Logger.msg(5, "JooqClusterStorage.getClusterContents() - uuid:"+uuid+" cluster:"+cluster+" primaryKeys"+Arrays.toString(primaryKeys));
+
+            return handler.getNextPrimaryKeys(JooqHandler.connect(), uuid, primaryKeys);
+        }
+        else
+            throw new PersistencyException("No handler found for cluster:'"+cluster+"'");
+    }
+
+    @Override
+    public C2KLocalObject get(ItemPath itemPath, String path) throws PersistencyException {
+        UUID uuid = itemPath.getUUID();
+
+        String[]    pathArray   = path.split("/");
+        String[]    primaryKeys = Arrays.copyOfRange(pathArray, 1, pathArray.length);
+        ClusterType cluster     = ClusterType.getValue(pathArray[0]);
+
+        JooqHandler handler = jooqHandlers.get(cluster);
+
+        if (handler != null) {
+            Logger.msg(5, "JooqClusterStorage.get() - uuid:"+uuid+" cluster:"+cluster+" primaryKeys:"+Arrays.toString(primaryKeys));
+
+            C2KLocalObject obj = handler.fetch(JooqHandler.connect(), uuid, primaryKeys);
+
+            if (obj == null && Logger.doLog(8)) {
+                Logger.warning(("JooqClusterStorage.get() - Could NOT fetch '"+itemPath+"/"+path+"'"));
+            }
+            return obj;
+        }
+        else
+            throw new PersistencyException("No handler found for cluster:'"+cluster+"'");
+    }
+
+    @Override
+    public void put(ItemPath itemPath, C2KLocalObject obj) throws PersistencyException {
+        put(itemPath, obj, null);
+    }
+
+    @Override
+    public void put(ItemPath itemPath, C2KLocalObject obj, Object locker) throws PersistencyException {
+        UUID uuid = itemPath.getUUID();
+        ClusterType cluster = obj.getClusterType();
+
+        JooqHandler handler = jooqHandlers.get(cluster);
+        
+         DSLContext dsl =  null;

```

Comment:
why is it called `dsl`? everywhere else it was called `context`.

Reply:
Changes will be on next commit.